Handly dynamic null checks against literal null

gafter opened the pull request:
Fixes #30939

gafter commented:
@dotnet/roslyn-compiler May I please have a couple of reviews of this nullable fix?

AlekseyTs commented:
Is there a typo in the title?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5149,7 +5149,27 @@ public override BoundNode VisitUnaryOperator(BoundUnaryOperator node)
+                case UnaryOperatorKind.DynamicFalse:
+                    Visit(node.Operand);
+                    Split();
+                    SetConditionalState(StateWhenFalse, StateWhenTrue);
```
>SetConditionalState(StateWhenFalse, StateWhenTrue); [](start = 20, length = 51)

Instead of always splitting the state, can we do this conditionally, only when we are in a conditional state (```IsConditionalState```)?

gafter replied:
As a rule we have avoided such micro-optimizations that complicate the code for little expected benefit.  The caller is likely to split the state in any case.

---
In reply to: [276385502](https://github.com/dotnet/roslyn/pull/34996#discussion_r276385502) [](ancestors = 276385502)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5149,7 +5149,27 @@ public override BoundNode VisitUnaryOperator(BoundUnaryOperator node)
             Debug.Assert(!IsConditionalState);
 
-            _ = base.VisitUnaryOperator(node);
+            switch (node.OperatorKind)
```
>switch (node.OperatorKind) [](start = 12, length = 26)

Should we move this more accurate analysis into the base class?

gafter replied:
The base class has the invariant that conditional states are only for `bool` values.  That is relaxed here.

---
In reply to: [276386205](https://github.com/dotnet/roslyn/pull/34996#discussion_r276386205) [](ancestors = 276386205)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5149,7 +5149,27 @@ public override BoundNode VisitUnaryOperator(BoundUnaryOperator node)
+                    // Yet we want to keep the result split if it was split.  So we simply visit.
+                    Visit(node.Operand);
+                    break;
+                case UnaryOperatorKind.DynamicFalse:
```
>UnaryOperatorKind.DynamicFalse [](start = 21, length = 30)

It feels like this case isn't covered by added tests.

gafter replied:
Agreed; it would not be particularly useful to do so, since it is only used for the implementation of a user-defined meaning for `||`, for which no reasonable inference can be produced by the caller.  Removing this case.

---
In reply to: [276389190](https://github.com/dotnet/roslyn/pull/34996#discussion_r276389190) [](ancestors = 276389190)

AlekseyTs commented:
Done with review pass (iteration 1)

gafter commented:
I am not going to fix the title because that would mess up threading of the conversation in the email notifications in outlook.

---
In reply to: [484211934](https://github.com/dotnet/roslyn/pull/34996#issuecomment-484211934) [](ancestors = 484211934)

gafter commented:
@AlekseyTs I think I have addressed your comments.  Do you have any others?

@dotnet/roslyn-compiler May I please have a second review for this small change?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -72989,28 +72989,249 @@ public void NullabilityOfTypeParameters_023()
-    void M<T>(dynamic? x)
+    void M1<T>(dynamic? x)
     {
         if (x == null) return;
```
>x == null [](start = 12, length = 9)

Consider adding tests for ```!(x != null)```

gafter replied:
Sure, but since we don't know the behavior, we don't infer anything.

---
In reply to: [278750198](https://github.com/dotnet/roslyn/pull/34996#discussion_r278750198) [](ancestors = 278750198)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -72989,28 +72989,249 @@ public void NullabilityOfTypeParameters_023()
     }
+    void M2<T>(dynamic? x)
+    {
+        if (x != null) return;
```
>x != null [](start = 12, length = 9)

Consider adding tests for ```!(x == null)```

gafter commented:
@AlekseyTs You may want to have another look.  I added support for dynamic `!`.

gafter commented:
@dotnet/roslyn-compiler May I please have a second review?
